Add schema validation for our settings file (#2530)
cobbler/cobbler: docs/cobbler-conf.rst

Cobbler Configuration
*********************

There are two main settings files which are located per default at ``/etc/cobbler/``:

- The file ``settings.yaml`` is following `YAML <https://yaml.org/spec/1.2/spec.html>`_ specification.
- The file ``modules.conf`` is following
  `INI <https://docs.python.org/3/library/configparser.html#supported-ini-file-structure>`_ specification.

.. note:: Since we are cleaning a lot of tech-debt this may change over time. We are trying to find the balance which
          format is the best for us to handle in the code and the best for admins to handle in the config files.

.. warning:: If you are using ``allow_dynamic_settings``, then the comments in the YAML file will vanish after the first
             change due to the fact that PyYAML doesn't support comments
             (`Source <https://github.com/yaml/pyyaml/issues/90>`_)

> There are additional configuration files location which need to follow the YAML Syntax. These are loaded from the

Reviewer: ```suggestion
There are additional configuration file locations which need to follow the YAML Syntax. These are loaded from the
```

``include`` directory in the ``settings.yaml`` file. Any key specified in one of these files overwrites values from the
main file.

.. warning:: When using ``allow_dynamic_settings`` the values are only persisted in the file ``settings.yaml``. This
             may lead to a non expected behaviour after ``cobblerd`` restarts. This is a
             `known issue <https://github.com/cobbler/cobbler/issues/2549>`_.

Updates to the yaml-settings-file
#################################

Starting with 3.2.1:

- We require the extension ``.yaml`` on our settings file to indicate the format of the file to editors and comply to
  standards of the YAML specification.

> - We require the usage of Booleans in the format of ``True`` and ``False``. If you have old integer style booleans with
>   ``1`` and ``0`` this is fine but may should convert them as soon as possible. We may decide in a future version to

Reviewer: ```suggestion
- We require the usage of booleans in the format of ``True`` and ``False``. If you have old integer style booleans with
  ``1`` and ``0`` this is fine but you may should convert them as soon as possible. We may decide in a future version to
```

  enforce our new way in a stricter manner. Automatic conversion is only done on a best-effort/available-resources
  basis.
- We enforce the types of values to the keys. Additional unexpected keys will throw errors. If you have those used in

>   Cobbler please report this in our Issue Tracker. We have decided to go this way to be able to rely on the existence

Reviewer: ```suggestion
  Cobbler please report this in our issue tracker. We have decided to go this way to be able to rely on the existence
```

  of the values. This gives us the freedom to write less access checks to the settings without loosing stability.

``settings.yaml``
#################

allow_duplicate_hostnames
=========================

If ``True``, Cobbler will allow insertions of system records that duplicate the ``--dns-name`` information of other
system records. In general, this is undesirable and should be left False.

default: ``False``

allow_duplicate_ips
===================

> If "True", Cobbler will allow insertions of system records that duplicate the IP address information of other system

Reviewer: ```suggestion
If ``True``, Cobbler will allow insertions of system records that duplicate the IP address information of other system
```

autoinstall_snippets_dir
========================

This is a directory of files that Cobbler uses to make templating easier. See the Wiki for more information. Changing
this directory should not be required.

default: ``/var/lib/cobbler/snippets``

autoinstall_templates_dir
=========================

This is a directory of files that Cobbler uses to make templating easier. See the Wiki for more information. Changing
this directory should not be required.

default: ``/var/lib/cobbler/templates``

bind_chroot_path
================

Set to path of bind chroot to create bind-chroot compatible bind configuration files. This should be automatically
detected.

default: ``""``

bind_master
===========

Set to the ip address of the master bind DNS server for creating secondary bind configuration files.

default: ``127.0.0.1``

boot_loader_conf_template_dir
=============================

Location of templates used for boot loader config generation.

default: ``"/etc/cobbler/boot_loader_conf"``

bootloaders_dir
===============

The location where Cobbler searches for the bootloaders to copy into the web directory.

default: ``/var/lib/cobbler/loaders``

grubconfig_dir
==============

> The location where Cobbler searches for Grub configuration files.

Reviewer: ```suggestion
The location where Cobbler searches for GRUB configuration files.
```

default: ``/var/lib/cobbler/grub_config``

build_reporting_*
=================

Email out a report when Cobbler finishes installing a system.

> - enabled: set to ``True`` to turn this feature on

Reviewer: ```suggestion
- enabled: Set to ``True`` to turn this feature on
```

- email: Which addresses to email
- ignorelist: TODO

> - sender: optional

Reviewer: ```suggestion
- sender: Optional
```

- smtp_server: Used to specify another server for an MTA.
- subject: Use the default subject unless overridden.

defaults:

.. code:: YAML

>     build_reporting_enabled: False

Reviewer: ```suggestion
    build_reporting_enabled: false
```

    build_reporting_sender: ""
    build_reporting_email: [ 'root@localhost' ]
    build_reporting_smtp_server: "localhost"
    build_reporting_subject: ""
    build_reporting_ignorelist: [ "" ]

buildisodir
===========

> Used for caching the intermediate files for ISO-Building. You may want to use an SSD, a tmpfs or something which is not
> persisted across reboots and can be easily thrown away but is also fast.

Reviewer: ```suggestion
Used for caching the intermediate files for ISO-Building. You may want to use a SSD, a tmpfs or something which does not
persist across reboots and can be easily thrown away but is also fast.
```

calls. If you will use Cobbler with config management or infrastructure-as-code tools such as Terraform, it is
recommended to disable by setting to ``False``.

default: ``True``

cheetah_import_whitelist
========================

Cheetah-language autoinstall templates can import Python modules. while this is a useful feature, it is not safe to
allow them to import anything they want. This whitelists which modules can be imported through Cheetah. Users can expand
this as needed but should never allow modules such as subprocess or those that allow access to the filesystem as Cheetah
templates are evaluated by ``cobblerd`` as code.

default:
 - ``random``
 - ``re``
 - ``time``
 - ``netaddr``

client_use_https
================

If set to ``True``, all commands to the API (not directly to the XML-RPC server) will go over HTTPS instead of plain
text. Be sure to change the ``http_port`` setting to the correct value for the web server.

default: ``False``

client_use_localhost
====================

If set to ``True``, all commands will be forced to use the localhost address instead of using the above value which can

> force commands like Cobbler sync to open a connection to a remote address if one is in the configuration and would

Reviewer: ```suggestion
force commands like ``cobbler sync`` to open a connection to a remote address if one is in the configuration and would
```

traceback.

default: ``False``

cobbler_master
==============

Used for replicating the Cobbler instance.

default: ``""``

convert_server_to_ip
====================

> Convert hostnames to IP-addresses (where possible) so DNS isn't a requirement for various tasks to work correctly.

Reviewer: ```suggestion
Convert hostnames to IP addresses (where possible) so DNS isn't a requirement for various tasks to work correctly.
```

default_template_type
=====================

The default template type to use in the absence of any other detected template. If you do not specify the template
with ``#template=<template_type>`` on the first line of your templates/snippets, Cobbler will assume try to use the
following template engine to parse the templates.

.. note:: Over time we will try to deprecate and remove Cheetah3 as a template engine. It is hard to package and there
          are fewer guides then with Jinja2. Making the templating independent of the engine is a task which complicates

>           the code. Thus please try to use Jinja2. We will try to support a seamless transition on a best-effort basis.

Reviewer: ```suggestion
          the code. Thus, please try to use Jinja2. We will try to support a seamless transition on a best-effort basis.
```

    ldap_server: "ldap.example.com"
    ldap_base_dn: "DC=example,DC=com"
    ldap_port: 389

>     ldap_tls: True
>     ldap_anonymous_bind: True

Reviewer: ```suggestion
    ldap_tls: true
    ldap_anonymous_bind: true
```

    ldap_search_bind_dn: ''
    ldap_search_passwd: ''
    ldap_search_prefix: 'uid='
    ldap_tls_cacertfile: ''
    ldap_tls_keyfile: ''
    ldap_tls_certfile: ''

bind_manage_ipmi
================

> When using the Bind9 DNS Server, you can enable or disable if the BMCs should receive own DNS entries.

Reviewer: ```suggestion
When using the Bind9 DNS server, you can enable or disable if the BMCs should receive own DNS entries.
```

default: ``False``

manage_dhcp
===========

Set to ``True`` to enable Cobbler's DHCP management features. The choice of DHCP management engine is in
``/etc/cobbler/modules.conf``.

default: ``True``

manage_dns
==========

Set to ``True`` to enable Cobbler's DNS management features. The choice of DNS management engine is in
``/etc/cobbler/modules.conf``.

default: ``False``

manage_*_zones
==============

If using BIND (named) for DNS management in ``/etc/cobbler/modules.conf`` and ``manage_dns`` is enabled (above), this
lists which zones are managed. See :ref:`dns-management` for more information.

defaults:

.. code::

    manage_forward_zones: []
    manage_reverse_zones: []

manage_genders
==============

> Whether or not to manage the Genders file. For more information on that visit:

Reviewer: ```suggestion
Whether or not to manage the genders file. For more information on that visit:
```

`github.com/chaos/genders <https://github.com/chaos/genders>`_

default: ``False``

manage_rsync
============

Set to ``True`` to enable Cobbler's RSYNC management features.

default: ``False``

manage_tftpd
==============

Set to ``True`` to enable Cobbler's TFTP management features. The choice of TFTP management engine is in
``/etc/cobbler/modules.conf``.

default: ``True``

mgmt_*
======

Cobbler has a feature that allows for integration with config management systems such as Puppet. The following
parameters work in conjunction with ``--mgmt-classes`` and are described in further detail at
:ref:`configuration-management`.

> .. code-block:: Yaml
> 
>     mgmt_classes: []
>     mgmt_parameters:
>         from_cobbler: True

Reviewer: ```suggestion
.. code-block:: YAML

    mgmt_classes: []
    mgmt_parameters:
        from_cobbler: true
```

next_server
===========

If using Cobbler with ``manage_dhcp``, put the IP address of the Cobbler server here so that PXE booting guests can find
it. If you do not set this correctly, this will be manifested in TFTP open timeouts.

default: ``127.0.0.1``

nsupdate_enabled
================

> This enabled or disables the replacement (or removal) of records in the DNS zone for systems created (or removed) by

Reviewer: ```suggestion
This enables or disables the replacement (or removal) of records in the DNS zone for systems created (or removed) by
```

leave this disabled. See the manpage for ``--netboot-enabled``.

default: ``True``

nopxe_with_triggers
===================

If this setting is set to ``True``, triggers will be executed when systems will request to toggle the
``--netboot-enabled`` record at the end of their installation.

default: ``True``

redhat_management_permissive
============================

If using ``authn_spacewalk`` in ``modules.conf`` to let Cobbler authenticate against Satellite/Spacewalk's auth system,
by default it will not allow per user access into Cobbler Web and Cobbler XML-RPC. In order to permit this, the following
setting must be enabled HOWEVER doing so will permit all Spacewalk/Satellite users of certain types to edit all of
Cobbler's configuration. these roles are: ``config_admin`` and ``org_admin``. Users should turn this on only if they
want this behavior and do not have a cross-multi-org separation concern. If you have a single org in your satellite,
it's probably safe to turn this on and then you can use CobblerWeb alongside a Satellite install.

default: ``False``

redhat_management_server
========================

> This setting is only used by the code that supports using Spacewalk/Satellite authentication within Cobbler Web and

Reviewer: ```suggestion
This setting is only used by the code that supports using Uyuni/SUSE Manager/Spacewalk/Satellite authentication within Cobbler Web and
```

> .. code::

Reviewer: ```suggestion
.. code:: YAML
```

>     restart_dns: True
>     restart_dhcp: True

Reviewer: ```suggestion
    restart_dns: true
    restart_dhcp: true
```

run_install_triggers
====================

Install triggers are scripts in ``/var/lib/cobbler/triggers/install`` that are triggered in autoinstall pre and post
sections. Any executable script in those directories is run. They can be used to send email or perform other actions.
They are currently run as root so if you do not need this functionality you can disable it, though this will also
disable ``cobbler status`` which uses a logging trigger to audit install progress.

> default: ``True``

Reviewer: ```suggestion
default: ``true``
```

> .. code::

Reviewer: ```suggestion
.. code:: YAML
```

>     scm_track_enabled: False

Reviewer: ```suggestion
    scm_track_enabled: false
```

    scm_track_mode: "git"
    scm_track_author: "cobbler <cobbler@localhost>"
    scm_push_script: "/bin/true"

serializer_pretty_json
======================

Sort and indent JSON output to make it more human-readable.

default: ``False``

server
======

This is the address of the Cobbler server -- as it is used by systems during the install process, it must be the address
or hostname of the system as those systems can see the server. if you have a server that appears differently to
different subnets (dual homed, etc), you need to read the ``--server-override`` section of the manpage for how that
works.

default: ``127.0.0.1``

sign_puppet_certs_automatically
===============================

When puppet starts on a system after installation it needs to have its certificate signed by the puppet master server.
Enabling the following feature will ensure that the puppet server signs the certificate after installation if the puppet
master server is running on the same machine as Cobbler. This requires ``puppet_auto_setup`` above to be enabled.

> default: ``False``

Reviewer: ```suggestion
default: ``false``
```

signature_path
==============

> The Cobbler Import workflow is powered by this file. Its location can be set with this config option.

Reviewer: ```suggestion
The Cobbler import workflow is powered by this file. Its location can be set with this config option.
```

default: ``/var/lib/cobbler/distro_signatures.json``

signature_url
=============

Updates to the signatures may happen more often then we have releases. To enable you to import new version we provide
the most up to date signatures we offer on this like. You may host this file for yourself and adjust it for your needs.

default: ``https://cobbler.github.io/signatures/3.0.x/latest.json``

tftpboot_location
=================

This variable contains the location of the tftpboot directory. If this directory is not present Cobbler does not start.

Default: ``/srv/tftpboot``

virt_auto_boot
==============

Should new profiles for virtual machines default to auto booting with the physical host when the physical host reboots?
This can be overridden on each profile or system object.

> default: ``True``

Reviewer: ```suggestion
default: ``true``
```

xmlrpc_port
===========

Cobbler's public XML-RPC listens on this port. Change this only if absolutely needed, as you'll have to start supplying
a new port option to Koan if it is not the default.

default: ``25151``

yum_distro_priority
===================

The default yum priority for all the distros. This is only used if yum-priorities plugin is used. 1 is the maximum
value. Tweak with caution.

> default: ``True``

Reviewer: ```suggestion
default: ``true``
```

yum_post_install_mirror
=======================

``cobbler repo add`` commands set Cobbler up with repository information that can be used during autoinstall and is
automatically set up in the Cobbler autoinstall templates. By default, these are only available at install time. To
make these repositories usable on installed systems (since Cobbler makes a very convenient mirror) set this to ``True``.
Most users can safely set this to ``True``. Users who have a dual homed Cobbler server, or are installing laptops that
will not always have access to the Cobbler server may wish to leave this as ``False``. In that case, the Cobbler

> mirrored yum repos are still accessible at ``http://cobbler.example.org/cblr/repo_mirror`` and yum configuration can

Reviewer: ```suggestion
mirrored yum repos are still accessible at ``http://cobbler.example.org/cblr/repo_mirror`` and YUM configuration can
```

tftpd
=====

> Chooses the TFTP management engine if ``manage_tftpd`` is enabled in ``/etc/cobbler/settings.yaml``, which is ON by

Reviewer: ```suggestion
Chooses the TFTP management engine if ``manage_tftpd`` is enabled in ``/etc/cobbler/settings.yaml``, which is **on** by
```
